### PR TITLE
[JUJU-768] LXD Container already exists

### PR DESCRIPTION
**Pull request by SimonRichardson**
If the container already exists when attempting to create a new
container, check that it matches the existing specs (profiles, devices
and config), if it does, return that container. Otherwise, return the
original error message.

## QA steps

See: https://bugs.launchpad.net/juju/+bug/1945813

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu --to lxd
$ juju add-unit ubuntu -n 5
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1945813

### CODE REVIEW

File: container/lxd/container.go [hunk 1]
```diff
@@ -262,11 +264,30 @@ func (s *Server) CreateContainerFromSpec(spec ContainerSpec) (*Container, error)
...
 	if err != nil {
+		if IsLXDAlreadyExists(err) {
+			container, _, conErr := s.GetContainer(spec.Name)
+			if conErr != nil {
+				return nil, errors.Trace(conErr)
+			}
+
+			// If we don't match the spec from the container, then we're not
+			// sure what we've got here. Return the original error message.
+			if container.Architecture != spec.Architecture ||
```
Comment (manadart) [inline container/lxd/container.go:280]: Should we also check that the container is in the running/started state?

I know that this solution corresponds with that suggested by @jameinel  on Launchpad, but my concern is that if there is some issue with an aborted creation on the LXD side, we can represent success without a usable instance.

File: container/lxd/container.go [hunk 2]
```diff
@@ -262,11 +264,40 @@ func (s *Server) CreateContainerFromSpec(spec ContainerSpec) (*Container, error)
...
+
+			switch container.StatusCode {
+			case api.Running:
+				c := Container{*container}
+				return &c, nil
+			case api.Started, api.Starting, api.Success:
+				// If the container isn't running, then send an error to try
+				// and do a retry. Unfortunately, there is no way to wait for
+				// the status of on a running containers.
+				return nil, errors.Errorf("waiting for container %s to be running, got: %d", req.Name, container.StatusCode)
```
Comment (hmlanigan) [inline container/lxd/container.go:296]: We should find out how this works in a multi-region lxd scenario.  Does the provisioner do what we think.

File: container/lxd/container.go (hunk 1 above)
Comment (SimonRichardson) [reply container/lxd/container.go:280]: There isn't a way to wait for the status to change, so we'll get the provisioner task to bounce it and retry to see if it's running. Alternatively, we could implement a wait timeout ?

File: container/lxd/container.go [hunk 3]
```diff
@@ -298,6 +312,50 @@ func (s *Server) CreateContainerFromSpec(spec ContainerSpec) (*Container, error)
 	return &c, nil
 }
 
+func (s *Server) waitForContainerToRunningState(spec ContainerSpec, ephemeral bool) (*api.Container, error) {
```
Comment (manadart) [inline container/lxd/container.go:315]: Bit mealy-mouthed. How about just `waitForRunningContainer`?

**Review by manadart — APPROVED** [review]
Is good.

**Pull request merged**